e2e/libs/keywords/statefulset_keywords.py

```diff
@@ -0,0 +1,141 @@
+from utility.constant import KIND_STATEFULSET
+from utility.constant import KIND_VOLUME
+
+
+class statefulset_keywords:
+
+    def __init__(self):
+        self.volume = Volume()
+        self.persistentvolumeclaim = PersistentVolumeClaim()
+
+    def cleanup_statefulsets(self):
+        cached = ResourceCache.get_kind(KIND_STATEFULSET)
+        if cached is None:
+            return
+
+        uncache = []
+        for name, links in cached.items():
+            delete_statefulset(name)
+            uncache.append(name)
+
+            if KIND_PERSISTENTVOLUMECLAIM in links:
+                pvc_name = links[KIND_PERSISTENTVOLUMECLAIM]
+                delete_persistentvolumeclaim(pvc_name)
+                ResourceCache.delete(KIND_PERSISTENTVOLUMECLAIM, pvc_name)
+
+        for name in uncache:
+            ResourceCache.delete(KIND_STATEFULSET, name)
+        assert len(ResourceCache.get_kind(KIND_STATEFULSET)) == 0
+
+    def create_statefulset(self, name, volume_type="rwo", option=""):
+        create_statefulset(name, volume_type, option)
+        ResourceCache.new(KIND_STATEFULSET, name)
+
+        pod_name = get_workload_pod_names(name)[0]
+        ResourceCache.new(KIND_POD, pod_name)
+        ResourceCache.create_or_update_link(
+            KIND_STATEFULSET, name, KIND_POD, pod_name
+        )
+
+        volume_name = get_workload_volume_name(name)
+        ResourceCache.new(KIND_VOLUME, volume_name)
+        ResourceCache.create_or_update_link(
+            KIND_STATEFULSET, name, KIND_VOLUME, volume_name
+        )
+
+        pvc_name = get_workload_pvc_name(name)
+        ResourceCache.new(KIND_PERSISTENTVOLUMECLAIM, pvc_name)
+        ResourceCache.create_or_update_link(
+            KIND_STATEFULSET, name, KIND_PERSISTENTVOLUMECLAIM, pvc_name
+        )
+
+    def expand_statefulset_claim_size_by_mib(self, statefulset_name, size_in_mib):
+        pvc_name = ResourceCache.get_link(KIND_STATEFULSET, statefulset_name, KIND_PERSISTENTVOLUMECLAIM)
+
+        size_in_byte = int(size_in_mib) * MEBIBYTE
+        expanded_size = self.persistentvolumeclaim.expand(pvc_name, size_in_byte)
+
+        ResourceCache.create_or_update_link(
+            KIND_STATEFULSET, statefulset_name, CACHE_EXPANDED_SIZE, expanded_size
+        )
+
+    def get_statefulset(self, statefulset_name):
+        return get_statefulset(statefulset_name)
+
+    def scale_statefulset(self, statefulset_name, replica_count):
+        return scale_statefulset(statefulset_name, replica_count)
+
+    def scale_statefulset_down(self, statefulset_name):
+        statefulset = self.get_statefulset(statefulset_name)
+        ResourceCache.create_or_update_link(
+            KIND_STATEFULSET, statefulset_name, CACHE_ORIGINAL_REPLICA_COUNT, statefulset.spec.replicas
+        )
+
+        scale_statefulset(statefulset_name, 0)
+
+        self.wait_for_statefulset_volume_detached(statefulset_name)
+
+    def scale_statefulset_back(self, statefulset_name):
```

Comment:
Can you help check these 2 failed tests:
https://ci.longhorn.io/job/private/job/longhorn-e2e-test/124/robot/report/log.html#s1-s1
https://ci.longhorn.io/job/private/job/longhorn-e2e-test/125/robot/report/log.html#s1-s1-t4
to see if there is a real issue or it's timeout issue? Thank you!

Reply:
If the tests failed in a run that previously succeeded in the iteration, it could be due to either flaky or a legitimate issue, this needs investigation. However, this failure isn't caused by this PR, so I suggest having a separate issue to track. This way, we can continue reviewing this PR, thank you.

Reply:
Fixed in https://github.com/longhorn/longhorn-tests/pull/1693.

Reply:
Related ticket: https://github.com/longhorn/longhorn/issues/7707